neural_lam/datastore/base.py

```diff
@@ -0,0 +1,371 @@
+        `(time)` if `is_forecast` is False.
+
+        If the data is ensemble data, the dataarray is expected to have an
+        additional `ensemble_member` dimension.
+
+        Parameters
+        ----------
+        category : str
+            The category of the dataset (state/forcing/static).
+        split : str
+            The time split to filter the dataset (train/val/test).
+
+        Returns
+        -------
+        xr.DataArray or None
+            The xarray DataArray object with processed dataset.
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def boundary_mask(self) -> xr.DataArray:
+        """
+        Return the boundary mask for the dataset, with spatial dimensions
+        stacked. Where the value is 1, the grid point is a boundary point, and
+        where the value is 0, the grid point is not a boundary point.
+
+        Returns
+        -------
+        xr.DataArray
+            The boundary mask for the dataset, with dimensions
+            `('grid_index',)`.
+
+        """
+        pass
+
+
+@dataclasses.dataclass
+class CartesianGridShape:
+    """Dataclass to store the shape of a grid."""
+
+    x: int
+    y: int
+
+
+class BaseCartesianDatastore(BaseDatastore):
+    """Base class for weather
+    data stored on a Cartesian
+    grid. In addition to the
+    methods and attributes
+    required for weather data
+    in general (see
+    `BaseDatastore`) for
+    Cartesian gridded source
+    data each `grid_index`
+    coordinate value is assume
+    to have an associated `x`
+    and `y`-value so that the
+    processed data-arrays can
+    be reshaped back into into
+    2D xy-gridded arrays.
+
+    In addition the following attributes and methods are required:
+    - `coords_projection` (property): Projection object for the coordinates.
+    - `grid_shape_state` (property): Shape of the grid for the state variables.
+    - `get_xy_extent` (method): Return the extent of the x, y coordinates for a
+      given category of data.
+    - `get_xy` (method): Return the x, y coordinates of the dataset.
+
+    """
+
+    CARTESIAN_COORDS = ["y", "x"]
+
+    @property
+    @abc.abstractmethod
+    def coords_projection(self) -> ccrs.Projection:
+        """Return the projection object for the coordinates.
+
+        The projection object is used to plot the coordinates on a map.
+
+        Returns
+        -------
+        cartopy.crs.Projection:
+            The projection object.
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def grid_shape_state(self) -> CartesianGridShape:
+        """The shape of the grid for the state variables.
+
+        Returns
+        -------
+        CartesianGridShape:
+            The shape of the grid for the state variables, which has `x` and
+            `y` attributes.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_xy(self, category: str, stacked: bool) -> np.ndarray:
```

Comment:
Is there a specific reason why this returns a `np.ndarray` rather than an `xr.Datarray/xr.Dataset`? That would maybe be more consistent with the rest of the methods.

Reply:
I left this returning `np.ndarray` because that is what the plotting functions use. I would prefer to change this in future, but I was trying to make as few modifications as possible. I got the impression from @sadamov that you @joeloskarsson felt like using plotting via xarray might be slower than working with numpy arrays directly. I would prefer to keep this as-is for now and we could improve this in a future PR

Reply:
Sure, we can handle it later. I was just thinking that you return an xarray here and then convert it to numpy in the plotting routine instead. No need to rework the plotting to use xarray just cause we change what the datastore returns.